posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,476 @@
+import json
+from collections.abc import Iterator
+from dataclasses import asdict, dataclass, field
+from datetime import datetime, timedelta, timezone
+from typing import NamedTuple
+from uuid import UUID
+
+import psycopg2
+from django.conf import settings
+from temporalio import activity, workflow
+
+from posthog.clickhouse.client.execute import sync_execute
+
+EPOCH = datetime(1970, 1, 1, 0, 0, tzinfo=timezone.utc)
+
+SELECT_PERSONS_TO_DELETE_QUERY = """
+SELECT
+    team_id,
+    old_person_id,
+    max(created_at) AS latest_created_at,
+    max(version) AS latest_version,
+    min(oldest_event) AS oldest_event_at
+FROM
+    {database}.person_overrides
+WHERE
+    created_at <= %(latest_created_at)s
+GROUP BY
+    team_id, old_person_id
+"""
+
+SELECT_LATEST_CREATED_AT_QUERY = """
+SELECT
+    max(created_at)
+FROM {database}.person_overrides;
+"""
+
+CREATE_JOIN_TABLE_QUERY = """
+CREATE OR REPLACE TABLE {database}.{table_name} (
+    team_id INT NOT NULL,
+    old_person_id UUID NOT NULL,
+    override_person_id UUID NOT NULL
+)
+ENGINE = Join(ANY, LEFT, team_id, old_person_id);
+"""
+
+INSERT_INTO_JOIN_TABLE_QUERY = """
+INSERT INTO {database}.{table_name}
+    SELECT
+        team_id,
+        old_person_id,
+        argMax(override_person_id, merged_at) AS override_person_id
+    FROM {database}.person_overrides
+    WHERE merged_at <= %(latest_created_at)s
+    GROUP BY team_id, old_person_id;
+"""
+
+SQUASH_EVENTS_QUERY = """
+ALTER TABLE
+    {database}.sharded_events
+UPDATE
+    person_id = joinGet({database}.{table_name}, 'override_person_id', toInt32(team_id), person_id)
+IN PARTITION
+    %(partition_id)s
+WHERE
+    true;
+"""
+
+DROP_JOIN_TABLE_QUERY = """
+DROP TABLE {database}.{table_name};
+"""
+
+DELETE_SQUASHED_PERSON_OVERRIDES_QUERY = """
+ALTER TABLE
+    {database}.person_overrides
+DELETE WHERE
+    team_id = %(team_id)s
+    AND old_person_id = %(old_person_id)s
+    AND merged_at <= %(latest_created_at)s;
+"""
+
+SELECT_CREATED_AT_FOR_PERSON_EVENT_QUERY = """
+SELECT
+    min(created_at) AS oldest_event_at
```

Comment:
shouldn't we use the event timestamp here? after-all the created at / oldest_event we had in PG was just a heuristic helper (which we use to optimise merge direction and based on our lookup here we might want to update PG with the real value)

Reply:
Similar argument also applies to user provided timestamps: they are not reliable.

Reply:
Ah I forgot we had ``created_at` DateTime64(6, 'UTC') DEFAULT now(),` in the schema my bad